File: 05-linked-list/reverse-linked-list/reverse_linked_list.py
"""
Reverse Linked List
=====================================================================
Read README.md first. Reverse a singly linked list and return the new
head. The ListNode class and test helpers are provided — you only
write reverse_list().

TODO(you): Restate the problem in ONE sentence:
    -

TODO(you): Approach — what three pointers (prev / current / next_temp)
           do you juggle, and in what order?
    -

--- Fill these in AFTER you solve it -------------------------------
Time complexity:  O(?)   #
Space complexity: O(?)   # iterative version — are you creating new nodes?
"""
from typing import List, Optional


class ListNode:
    """A single node in a singly linked list. (Provided — don't change.)"""
    def __init__(self, val: int = 0, next: "Optional[ListNode]" = None):
        self.val = val
        self.next = next


def reverse_list(head: Optional[ListNode]) -> Optional[ListNode]:
    # -----------------------------------------------------------------
    # TODO(you): Step 1 — start `prev = None`, `current = head`.
    # TODO(you): Step 2 — while current is not None:
    #                       (a) save next_temp = current.next
    #                       (b) point current.next back to prev
    #                       (c) advance prev = current, current = next_temp
    # TODO(you): Step 3 — the new head is `prev`. Return it.
    # -----------------------------------------------------------------
    prev = None
    current = head
    while current is not None:
        next_up = current.next
        current.next = prev
        prev = current
        current = next_up
    return prev


# ---- Test helpers (provided — don't change) ------------------------
def build_list(values: List[int]) -> Optional[ListNode]:
    head: Optional[ListNode] = None
    for v in reversed(values):
        head = ListNode(v, head)
    return head


def to_list(head: Optional[ListNode]) -> List[int]:
    out: List[int] = []
    while head is not None:
        out.append(head.val)
        head = head.next
    return out


# ---- Self-check tests (do NOT edit) --------------------------------
# Run:  python3 reverse_linked_list.py
if __name__ == "__main__":
    assert to_list(reverse_list(build_list([1, 2, 3, 4, 5]))) == [5, 4, 3, 2, 1]
    assert to_list(reverse_list(build_list([1, 2]))) == [2, 1]
    assert to_list(reverse_list(build_list([]))) == []
    assert to_list(reverse_list(build_list([7]))) == [7]
    print("✅ All tests passed!")

File: 05-linked-list/reverse-linked-list/reverseLinkedList.ts
/**
 * Reverse Linked List
 * ===================================================================
 * Read README.md first. Reverse a singly linked list and return the
 * new head. The ListNode class and test helpers are provided — you
 * only write reverseList().
 *
 * TODO(you): Restate the problem in ONE sentence:
 *   -
 *
 * TODO(you): Approach — what three pointers (prev / current / nextTemp)
 *            do you juggle, and in what order?
 *   -
 *
 * --- Fill these in AFTER you solve it -----------------------------
 * Time complexity:  O(?)   //
 * Space complexity: O(?)   // iterative version — creating new nodes?
 */

import assert from "node:assert";

// A single node in a singly linked list. (Provided — don't change.)
export class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val: number = 0, next: ListNode | null = null) {
    this.val = val;
    this.next = next;
  }
}

export function reverseList(head: ListNode | null): ListNode | null {
  // -------------------------------------------------------------------
  // TODO(you): Step 1 — start `prev = null`, `current = head`.
  // TODO(you): Step 2 — while current is not null:
  //                       (a) save nextTemp = current.next
  //                       (b) point current.next back to prev
  //                       (c) advance prev = current, current = nextTemp
  // TODO(you): Step 3 — the new head is `prev`. Return it.
  // -------------------------------------------------------------------
  let prev: ListNode | null = null;
  let current: ListNode | null = head;
    while (current != null) {
      let nextUp = current.next;
      current.next = prev;
      prev = current;
      current = nextUp;
    }
  return prev
}

// ---- Test helpers (provided — don't change) -------------------------
function buildList(values: number[]): ListNode | null {
  let head: ListNode | null = null;
  for (let i = values.length - 1; i >= 0; i--) {
    head = new ListNode(values[i]!, head);
  }
  return head;
}

function toArray(head: ListNode | null): number[] {
  const out: number[] = [];
  while (head !== null) {
    out.push(head.val);
    head = head.next;
  }
  return out;
}

// ---- Self-check tests (do NOT edit) ---------------------------------
// Run:  npx tsx reverseLinkedList.ts
assert.deepStrictEqual(toArray(reverseList(buildList([1, 2, 3, 4, 5]))), [5, 4, 3, 2, 1]);
assert.deepStrictEqual(toArray(reverseList(buildList([1, 2]))), [2, 1]);
assert.deepStrictEqual(toArray(reverseList(buildList([]))), []);
assert.deepStrictEqual(toArray(reverseList(buildList([7]))), [7]);
console.log("✅ All tests passed!");
